File: gunhoflash/reorder-list.py
from collections import deque
from typing import List, Optional

class ListNode:
  def __init__(self, val = 0, next = None):
    self.val = val
    self.next = next

class Solution:
  def reorderList(self, head: Optional[ListNode]) -> None:
    deq = deque()
    while head:
      deq.append(head)
      head = head.next

    tail = deq.popleft()
    while True:
      if deq:
        tail.next = deq.pop()
        tail = tail.next
      else:
        break
      if deq:
        tail.next = deq.popleft()
        tail = tail.next
      else:
        break
    tail.next = None

File: gunhoflash/reorder-list.ts
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function reorderList(head: ListNode | null): void {
  const deq: ListNode[] = [];
  while (head) {
    deq.push(head);
    head = head.next;
  }

  let tail = deq.shift()!;
  while (true) {
    tail.next = deq.pop() || null;
    if (tail.next === null) return;
    tail = tail.next;
    tail.next = deq.shift() || null;
    if (tail.next === null) return;
    tail = tail.next;
  }
}
